File: android/dotrino-native/src/main/java/com/dotrino/sdk/VaultClient.kt
package com.dotrino.sdk

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

/**
 * An account this phone approves for natively: which vault, through which proxy, and the
 * paper (`cert`) the vault gave to this device's key. No secret in here — the keys live in
 * the Keystore under [id].
 */
@Serializable
data class Account(
    val id: String,
    val name: String,
    val profileId: String? = null,
    val vault: String,
    val proxy: String,
    val cert: JsonObject,
    val deviceId: String,
    val addedAt: Long = System.currentTimeMillis(),
)

/** A pending request, as the vault lists it. [ctx] is already opened with my key, when it came sealed. */
data class Approval(
    val id: String,
    val ns: String,
    val kind: String,
    val deviceId: String,
    val label: String,
    val exp: Long,
    val ctx: JsonObject?,
    /** Why the context could not be read, when it came but did not open (or was not sealed to me). */
    val ctxError: String?,
)

data class Grant(val id: String, val ns: String, val deviceId: String, val label: String, val exp: Long, val uses: Long, val ctx: JsonObject?)

class VaultError(message: String, val code: String) : Exception(message)

/**
 * Talks to ONE account's vault over one [ProxyConnection]. Same messages as `vaultRpc` in
 * `@dotrino/identity/vault/remote.js`: `{ type, data: {…, publickey, ts}, signature, cert }`
 * to the vault's key, and the answer comes back to this connection.
 *
 * The vault's answers carry no request id, so calls on one account go ONE AT A TIME (the
 * mutex): with two in flight, the first answer would settle whichever asked first.
 */
class VaultClient(
    account: Account,
    private val keys: DeviceKeys,
    private val conn: ProxyConnection,
    /** Called with the renewed account so the caller persists the new paper. */
    private val onRenewed: (Account) -> Unit = {},
) {
    companion object {
        const val SECRETS = "vault.secrets"
        const val SECRETS_RESULT = "vault.secrets.result"
        const val RENEW = "vault.renew"
        const val RENEWED = "vault.renewed"
        const val ERROR = "vault.error"
        const val ADMIN_EVENT = "vault.admin.event"
        const val SCOPE_APPROVE = "vault:approve"
        private val json = Json { ignoreUnknownKeys = true }
    }

    @Volatile var account: Account = account; private set
    private val lock = Mutex()

    private suspend fun rpc(sendType: String, okType: String, data: JsonObject, timeoutMs: Long = 15_000): JsonObject = lock.withLock {
        val signed = JsonObject(data + mapOf("publickey" to JsonPrimitive(keys.publickey), "ts" to JsonPrimitive(System.currentTimeMillis())))
        val answer = CompletableDeferred<JsonObject>()
        val off = conn.onMessage { m ->
            when (m.payload["type"]?.jsonPrimitive?.content) {
                okType -> answer.complete(m.payload)
                ERROR -> {
                    val msg = m.payload["error"]?.jsonPrimitive?.content ?: "vault error"
                    answer.completeExceptionally(VaultError(msg, codeOf(msg)))
                }
            }
        }
        try {
            conn.sendByPubkey(account.vault, buildJsonObject {
                put("type", sendType); put("data", signed)
                put("signature", keys.sign(Canonical.stringify(signed))); put("cert", account.cert)
            })
            try { withTimeout(timeoutMs) { answer.await() } }
            catch (e: kotlinx.coroutines.TimeoutCancellationException) {
                throw VaultError("the vault did not reply (is it running?)", "vault-no-reply")
            }
        } finally { off() }
    }

    /** `unauthorized: <reason>` → the reason, which is what can be acted on. */
    private fun codeOf(msg: String): String = Regex("^unauthorized: ([\\w-]+)").find(msg)?.groupValues?.get(1) ?: "vault-error"

    /**
     * Once, and only for a paper problem: after `+aprueba` the old paper does not carry
     * `vault:approve`, and after a record change it is behind. A revoked device or one the
     * record no longer lets approve is NOT retried — renewing cannot fix that, and retrying
     * would only hide the reason.
     */
    private suspend fun <T> withPaper(block: suspend () -> T): T = try { block() } catch (e: VaultError) {
        if (e.code in setOf("scope", "acta-vieja", "untrusted-issuer", "expired", "legacy-cert-retirado", "no-acta")) { renew(); block() } else throw e
    }

    /** A new paper for my key. Accepted only if it is from the vault I enrolled with, for my key, and lets me approve. */
    suspend fun renew(): Account {
        val res = rpc(RENEW, RENEWED, buildJsonObject { put("op", "renew") })
        val cert = res["cert"] as? JsonObject ?: throw VaultError("the vault did not send a paper", "no-cert")
        Delegation.check(cert, account.vault, keys.publickey, SCOPE_APPROVE)?.let {
            throw VaultError("invalid renewed paper: $it", it)
        }
        account = account.copy(cert = cert)
        onRenewed(account)
        return account
    }

    private suspend fun secrets(data: JsonObject): JsonObject = withPaper {
        val res = rpc(SECRETS, SECRETS_RESULT, data)
        res["body"] as? JsonObject ?: throw VaultError("the vault answered without a body", "no-body")
    }

    suspend fun approvals(): List<Approval> {
        val body = secrets(buildJsonObject { put("op", "approvals") })
        return (body["items"] as? JsonArray ?: JsonArray(emptyList())).mapNotNull { (it as? JsonObject)?.let(::approvalOf) }
    }

    suspend fun approve(id: String) = answer("approve", id)
    suspend fun deny(id: String) = answer("deny", id)

    private suspend fun answer(op: String, id: String) {
        val body = secrets(buildJsonObject { put("op", op); put("id", id) })
        if (body["ok"]?.jsonPrimitive?.booleanOrNull != true) throw VaultError("the vault did not confirm the $op", "not-confirmed")
    }

    suspend fun grants(): List<Grant> {
        val body = secrets(buildJsonObject { put("op", "grants") })
        return (body["items"] as? JsonArray ?: JsonArray(emptyList())).mapNotNull { e ->
            val o = e as? JsonObject ?: return@mapNotNull null
            Grant(
                id = o.str("id") ?: return@mapNotNull null, ns = o.str("ns") ?: "", deviceId = o.str("deviceId") ?: "",
                label = o.str("label") ?: "", exp = o.long("exp") ?: 0, uses = o.long("uses") ?: 0, ctx = openCtx(o).first,
            )
        }
    }

    suspend fun revokeGrant(id: String) {
        val body = secrets(buildJsonObject { put("op", "grant-revoke"); put("id", id) })
        if (body["ok"]?.jsonPrimitive?.booleanOrNull != true) throw VaultError("that approval was no longer active", "not-found")
    }

    private fun approvalOf(o: JsonObject): Approval? {
        val (ctx, err) = openCtx(o)
        return Approval(
            id = o.str("id") ?: return null, ns = o.str("ns") ?: "", kind = o.str("kind") ?: "read",
            deviceId = o.str("deviceId") ?: "", label = o.str("label") ?: "", exp = o.long("exp") ?: 0,
            ctx = ctx, ctxError = err,
        )
    }

    /** The context comes sealed to MY encryption key; open it here. Not opening is said, never hidden. */
    private fun openCtx(o: JsonObject): Pair<JsonObject?, String?> {
        (o["ctx"] as? JsonObject)?.let { return it to null }
        val wrap = o["ctxWrap"] as? JsonObject
        val env = o["ctxEnvelope"] as? JsonObject
        if (wrap != null && env != null) {
            return try { (json.parseToJsonElement(Crypto.openSealed(wrap, env, keys)) as JsonObject) to null }
            catch (e: Exception) { null to "cannot-open" }
        }
        if (o["ctxSealed"]?.jsonPrimitive?.booleanOrNull == false) return null to (o.str("ctxReason") ?: "not-sealed")
        return null to null
    }

    private fun JsonObject.str(k: String) = (this[k] as? JsonPrimitive)?.takeIf { it.isString }?.content
    private fun JsonObject.long(k: String) = (this[k] as? JsonPrimitive)?.longOrNull
}

File: ios/DotrinoNative/VaultClient.swift
import Foundation

/// A pending request, as the vault lists it. [ctx] is already opened with my key, when it came sealed.
public struct Approval: Equatable, Sendable, Identifiable {
    public let id: String
    public let ns: String
    public let kind: String
    public let deviceId: String
    public let label: String
    public let exp: Int64
    public let ctx: JSON?
    /// Why the context could not be read, when it came but did not open (or was not sealed to me).
    public let ctxError: String?
}

public struct Grant: Equatable, Sendable, Identifiable {
    public let id: String
    public let ns: String
    public let deviceId: String
    public let label: String
    public let exp: Int64
    public let uses: Int64
    public let ctx: JSON?
}

public struct VaultError: Error, CustomStringConvertible {
    public let description: String
    public let code: String
    public init(_ d: String, code: String) { description = d; self.code = code }
}

/// Calls go one at a time (see [VaultClient]).
actor AsyncLock {
    private var busy = false
    private var waiters: [CheckedContinuation<Void, Never>] = []

    func lock() async {
        if !busy { busy = true; return }
        await withCheckedContinuation { waiters.append($0) }
    }

    func unlock() {
        if waiters.isEmpty { busy = false } else { waiters.removeFirst().resume() }
    }
}

/// Talks to ONE account's vault over one [ProxyConnection]. Same messages as `vaultRpc` in
/// `@dotrino/identity/vault/remote.js`: `{ type, data: {…, publickey, ts}, signature, cert }`
/// to the vault's key, and the answer comes back to this connection.
///
/// The vault's answers carry no request id, so calls on one account go ONE AT A TIME (the
/// lock): with two in flight, the first answer would settle whichever asked first.
public final class VaultClient: @unchecked Sendable {
    public static let secrets = "vault.secrets"
    public static let secretsResult = "vault.secrets.result"
    public static let renewType = "vault.renew"
    public static let renewed = "vault.renewed"
    public static let errorType = "vault.error"
    public static let adminEvent = "vault.admin.event"
    public static let scopeApprove = "vault:approve"
    /// The record does not let this key approve (yet): the owner has to give it `+aprueba`.
    public static let noApprove = "no-approve"

    private let keys: DeviceKeys
    private let conn: ProxyConnection
    private let onRenewed: (Account) -> Void
    private let gate = AsyncLock()
    private let stateLock = NSLock()
    private var _account: Account

    public var account: Account { stateLock.lock(); defer { stateLock.unlock() }; return _account }

    /// [onRenewed] is called with the renewed account so the caller persists the new paper.
    public init(account: Account, keys: DeviceKeys, conn: ProxyConnection, onRenewed: @escaping (Account) -> Void = { _ in }) {
        _account = account; self.keys = keys; self.conn = conn; self.onRenewed = onRenewed
    }

    private func rpc(_ sendType: String, _ okType: String, _ data: [String: JSON], timeout: TimeInterval = 15) async throws -> JSON {
        await gate.lock()
        do {
            let r = try await rpcUnlocked(sendType, okType, data, timeout: timeout)
            await gate.unlock()
            return r
        } catch {
            await gate.unlock()
            throw error
        }
    }

    private func rpcUnlocked(_ sendType: String, _ okType: String, _ data: [String: JSON], timeout: TimeInterval) async throws -> JSON {
        var d = data
        d["publickey"] = .string(keys.publickey)
        d["ts"] = .int(nowMs())
        let signed = JSON.object(d)
        let answer = OneShot<JSON>()
        let off = conn.onMessage { m in
            switch m.payload["type"]?.string {
            case okType: answer.finish(.success(m.payload))
            case VaultClient.errorType:
                let msg = m.payload["error"]?.string ?? "vault error"
                answer.finish(.failure(VaultError(msg, code: VaultClient.codeOf(msg))))
            default: break
            }
        }
        defer { off() }
        try conn.sendByPubkey(account.vault, ["type": .string(sendType), "data": signed,
                                              "signature": .string(keys.sign(Canonical.stringify(signed))), "cert": account.cert])
        return try await answer.wait(timeout: timeout, onTimeout: VaultError("the vault did not reply (is it running?)", code: "vault-no-reply"))
    }

    /// `unauthorized: <reason>` → the reason, which is what can be acted on.
    static func codeOf(_ msg: String) -> String {
        guard let r = msg.range(of: #"^unauthorized: ([\w-]+)"#, options: .regularExpression) else { return "vault-error" }
        return String(msg[r].dropFirst("unauthorized: ".count))
    }

    /// Once, and only for a paper problem: after `+aprueba` the old paper does not carry
    /// `vault:approve`, and after a record change it is behind. A revoked device or one the
    /// record no longer lets approve is NOT retried — renewing cannot fix that, and retrying
    /// would only hide the reason.
    private func withPaper<T>(_ block: () async throws -> T) async throws -> T {
        do { return try await block() } catch let e as VaultError
            where ["scope", "acta-vieja", "untrusted-issuer", "expired", "legacy-cert-retirado", "no-acta"].contains(e.code) {
            _ = try await renew()
            return try await block()
        }
    }

    /// A new paper for my key. Accepted only if it is from the vault I enrolled with, for my key, and lets me approve.
    @discardableResult
    public func renew() async throws -> Account {
        let res = try await rpc(Self.renewType, Self.renewed, ["op": "renew"])
        guard let cert = res["cert"], cert.object != nil else { throw VaultError("the vault did not send a paper", code: "no-cert") }
        let a = account
        if let why = Delegation.check(cert, vault: a.vault, sub: keys.publickey, expectedScope: Self.scopeApprove) {
            // A good paper without `vault:approve` is not a broken paper: the record simply
            // does not let this key approve yet (`+aprueba` is missing). Said as such, so the
            // screen can tell the owner what to do instead of «invalid renewed paper: scope».
            if why == "scope" { throw VaultError("the record does not let this device approve", code: Self.noApprove) }
            throw VaultError("invalid renewed paper: \(why)", code: why)
        }
        var n = a
        n.cert = cert
        stateLock.withLock { _account = n }
        onRenewed(n)
        return n
    }

    private func secretsCall(_ data: [String: JSON]) async throws -> JSON {
        try await withPaper {
            let res = try await rpc(Self.secrets, Self.secretsResult, data)
            guard let body = res["body"], body.object != nil else { throw VaultError("the vault answered without a body", code: "no-body") }
            return body
        }
    }

    public func approvals() async throws -> [Approval] {
        let body = try await secretsCall(["op": "approvals"])
        return (body["items"]?.array ?? []).compactMap(approvalOf)
    }

    public func approve(_ id: String) async throws { try await answer("approve", id) }
    public func deny(_ id: String) async throws { try await answer("deny", id) }

    private func answer(_ op: String, _ id: String) async throws {
        let body = try await secretsCall(["op": .string(op), "id": .string(id)])
        if body["ok"]?.bool != true { throw VaultError("the vault did not confirm the \(op)", code: "not-confirmed") }
    }

    public func grants() async throws -> [Grant] {
        let body = try await secretsCall(["op": "grants"])
        return (body["items"]?.array ?? []).compactMap { o in
            guard let id = o["id"]?.string else { return nil }
            return Grant(id: id, ns: o["ns"]?.string ?? "", deviceId: o["deviceId"]?.string ?? "", label: o["label"]?.string ?? "",
                         exp: o["exp"]?.int ?? 0, uses: o["uses"]?.int ?? 0, ctx: openCtx(o).0)
        }
    }

    public func revokeGrant(_ id: String) async throws {
        let body = try await secretsCall(["op": "grant-revoke", "id": .string(id)])
        if body["ok"]?.bool != true { throw VaultError("that approval was no longer active", code: "not-found") }
    }

    private func approvalOf(_ o: JSON) -> Approval? {
        guard let id = o["id"]?.string else { return nil }
        let (ctx, err) = openCtx(o)
        return Approval(id: id, ns: o["ns"]?.string ?? "", kind: o["kind"]?.string ?? "read", deviceId: o["deviceId"]?.string ?? "",
                        label: o["label"]?.string ?? "", exp: o["exp"]?.int ?? 0, ctx: ctx, ctxError: err)
    }

    /// The context comes sealed to MY encryption key; open it here. Not opening is said, never hidden.
    private func openCtx(_ o: JSON) -> (JSON?, String?) {
        if let c = o["ctx"], c.object != nil { return (c, nil) }
        if let wrap = o["ctxWrap"], wrap.object != nil, let env = o["ctxEnvelope"], env.object != nil {
            do {
                let c = try JSON.parse(Crypto.openSealed(wrap: wrap, envelope: env, keys: keys))
                return c.object != nil ? (c, nil) : (nil, "cannot-open")
            } catch { return (nil, "cannot-open") }
        }
        if o["ctxSealed"]?.bool == false { return (nil, o["ctxReason"]?.string ?? "not-sealed") }
        return (nil, nil)
    }
}
